src/main/java/generator/ResponseHydrator.java

```diff
@@ -0,0 +1,165 @@
+package generator;
+
+import graphql.Scalars;
+import graphql.language.*;
+import graphql.schema.*;
+import hydrator.IScalarHydrator;
+import hydrator.IValueHydratorFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Hydrates the response to a query/mutation.
+ */
+public class ResponseHydrator {
+    private final GraphQLSchema graphQLSchema;
+    private final IValueHydratorFactory valueHydratorFactory;
+
+    public ResponseHydrator(GraphQLSchema graphQLSchema, IValueHydratorFactory valueHydratorFactory) {
+        this.graphQLSchema = graphQLSchema;
+        this.valueHydratorFactory = valueHydratorFactory;
+    }
+
+    /**
+     * Hydrates a parsed tree using the valueHydratorFactory.
+     * @param document Parsed tree document.
+     * @return ResponseNode tree containing the hydrated response.
+     */
+    public ResponseNode hydrate(Document document) {
+        return hydrate(document, null);
+    }
+
+    /**
+     * Hydrates a parsed tree using the valueHydratorFactory and overrides values.
+     * @param document Parsed tree document.
+     * @param override Map containing values to override.
+     * @return ResponseNode tree containing the hydrated response.
+     */
+    public ResponseNode hydrate(Document document, Map<String, Object> override) {
+        Map<String, ResponseNode> children = new HashMap<>();
+
+        for (Definition definition : document.getDefinitions()) {
+            // Handle top-level query and mutation requests.
+            if (definition instanceof OperationDefinition) {
+                OperationDefinition operationDefinition = (OperationDefinition) definition;
+
+                // TODO (othebe): Check for query or mutation.
+                ResponseNode selectionResponse = hydrateSelectionRec(
+                        graphQLSchema.getQueryType(),
+                        operationDefinition.getSelectionSet(),
+                        override);
+
+                children.put(operationDefinition.getName(), selectionResponse);
+            } else {
+                throw new RuntimeException(String.format("Unhandled definition: %s", definition.toString()));
+            }
+        }
+
+        return ResponseNode.createBranchNode(children);
+    }
+
+    /**
+     * Helper function to hydrate a GraphQLType with fields provided in the selectionSet.
+     * @param parentType SelectionSet parent type.
+     * @param selectionSet Fields to include for the parent.
+     * @param override Map containing values to override.
+     * @return ResponseNode tree containing the hydrated response.
+     */
+    private ResponseNode hydrateSelectionRec(GraphQLOutputType parentType, SelectionSet selectionSet, Map<String, Object> override) {
```

Comment:
Holy moly, why are we doing all the parsing manually?? graphql-java already has methods to parse and respond to queries, doesn't it?

Reply:
This is related to your other comment: https://github.com/othebe/graphql-response-generator/pull/1#discussion_r144680160

I tried using the built-in lifeCycle of graphql-java as a first pass, but my experience from this was that our use-case deviated slightly from its its purpose, so there was a lot of logic that was really dependent on the internal workings. We should discuss this in person, but one example was how every interfaces requires a type resolver to a type, and inferring the type during runtime could be complex, and passing it in involves too much setup for the user, when in fact we just care about the fields.

I think https://github.com/graphql-java/graphql-java/blob/master/src/test/groovy/graphql/StarWarsQueryTest.groovy is the way you had in mind, but lot of schema knowledge and wiring required.